### PR TITLE
Remove IMessageSerializer, use IZeroMessageSerializer for all messages

### LINKED ISSUE #3508: Move IMessageSerializer to IZeroMessageSerializer
To avoid  allocations

### PR DESCRIPTION
**Pull request by tanishqjasoria**
Resolves #3508 

## Changes:
- remove IMessageSerializer
- port classes implementing IMessageSerializer to IZeroMessageSerializer
- Reduces allocations

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [x] Yes
- [ ] No

**In case you checked yes, did you write tests??**

- [ ] Yes
- [x] No

**Comments about testing , should you have some** (optional)
This was merged before once and caused memory leak.
https://github.com/NethermindEth/nethermind/pull/4292
https://github.com/NethermindEth/nethermind/pull/4425

## Further comments (optional)

This need to be tested properly 
- what performance gains are we getting
- the memory leak issue is resolved or not

### CODE REVIEW

**Comment by MarekM25** [comment]
Do we have any metrics/measurements how much we reduced memory allocations with this PR?

**Comment by MarekM25** [comment]
Not all tests are passing. Flaky tests? How the PR was tested?

File: src/Nethermind/Nethermind.Network/P2P/NettyRlpStream.cs [hunk 1]
```diff
@@ -97,8 +98,12 @@ public override int Position
 
         public override int Length => _buffer.ReadableBytes + (_buffer.ReaderIndex - _initialPosition);
 
-        public override bool HasBeenRead => _buffer.ReadableBytes > 0;
+        public override bool HasBeenRead => _buffer.ReadableBytes <= 0;
 
         protected override string Description => "|NettyRlpStream|description missing|";
+        public void Dispose()
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network/P2P/NettyRlpStream.cs:104]: Are we sure we are disposing it with every use?

File: src/Nethermind/Nethermind.Network/Rlpx/Handshake/Eip8MessagePad.cs [hunk 2]
```diff
@@ -33,5 +34,11 @@ public byte[] Pad(byte[] message)
             byte[] padding = _cryptoRandom.GenerateRandomBytes(100 + _cryptoRandom.NextInt(201));
             return Bytes.Concat(message, padding);
         }
+
+        public void Pad(IByteBuffer message)
+        {
+            byte[] padding = _cryptoRandom.GenerateRandomBytes(100 + _cryptoRandom.NextInt(201));
+            message.WriteBytes(padding);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network/Rlpx/Handshake/Eip8MessagePad.cs:41]: maybe make this padding in span on stack?

File: src/Nethermind/Nethermind.Network/Rlpx/Handshake/HandshakeService.cs [hunk 3]
```diff
@@ -67,33 +70,51 @@ public Packet Auth(PublicKey remoteNodeId, EncryptionHandshake handshake, bool p
...
+                    PublicKey = _privateKey.PublicKey,
+                    Signature = _ecdsa.Sign(handshake.EphemeralPrivateKey, new Keccak(forSigning))
+                };
+
+                IByteBuffer authData = _messageSerializationService.ZeroSerialize(authMessage);
+                try
+                {
+                    int size = authData.ReadableBytes + 32 + 16 + 65; // data + MAC + IV + pub
+                    byte[] sizeBytes = size.ToBigEndianByteArray().Slice(2, 2);
+                    byte[] packetData = _eciesCipher.Encrypt(remoteNodeId, authData.ReadAllBytesAsArray(), sizeBytes);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network/Rlpx/Handshake/HandshakeService.cs:109]: In the end we have ReadAllBytesAsArray(), maybe we can do it with span?

File: src/Nethermind/Nethermind.Network/Rlpx/Handshake/HandshakeService.cs [hunk 4]
```diff
@@ -67,33 +70,51 @@ public Packet Auth(PublicKey remoteNodeId, EncryptionHandshake handshake, bool p
...
+                    PublicKey = _privateKey.PublicKey,
+                    Signature = _ecdsa.Sign(handshake.EphemeralPrivateKey, new Keccak(forSigning)),
+                    IsTokenUsed = false,
+                    EphemeralPublicHash = Keccak.Compute(handshake.EphemeralPrivateKey.PublicKey.Bytes)
+                };
+
+                IByteBuffer authData = _messageSerializationService.ZeroSerialize(authMessage);
+                try
+                {
+                    byte[] packetData = _eciesCipher.Encrypt(remoteNodeId, authData.ReadAllBytesAsArray(), Array.Empty<byte>());
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network/Rlpx/Handshake/HandshakeService.cs:85]: In the end we have ReadAllBytesAsArray(), maybe we can do it with span?

File: src/Nethermind/Nethermind.Network/Rlpx/Handshake/HandshakeService.cs [hunk 5]
```diff
@@ -143,24 +164,41 @@ public Packet Ack(EncryptionHandshake handshake, Packet auth)
...
+                AckMessage ackMessage = new()
+                {
+                    EphemeralPublicKey = handshake.EphemeralPrivateKey.PublicKey,
+                    Nonce = handshake.RecipientNonce
+                };
+
+                IByteBuffer ackData = _messageSerializationService.ZeroSerialize(ackMessage);
+                try
+                {
+                    data = _eciesCipher.Encrypt(handshake.RemoteNodeId, ackData.ReadAllBytesAsArray(), Array.Empty<byte>());
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network/Rlpx/Handshake/HandshakeService.cs:176]: In the end we have ReadAllBytesAsArray(), maybe we can do it with span?

File: src/Nethermind/Nethermind.Network/Rlpx/Handshake/HandshakeService.cs [hunk 6]
```diff
@@ -143,24 +164,41 @@ public Packet Ack(EncryptionHandshake handshake, Packet auth)
...
+                {
+                    EphemeralPublicKey = handshake.EphemeralPrivateKey.PublicKey,
+                    Nonce = handshake.RecipientNonce
+                };
+                IByteBuffer ackData = _messageSerializationService.ZeroSerialize(ackMessage);
+                try
+                {
+                    int size = ackData.ReadableBytes + 32 + 16 + 65; // data + MAC + IV + pub
+                    byte[] sizeBytes = size.ToBigEndianByteArray().Slice(2, 2);
+                    data = Bytes.Concat(sizeBytes, _eciesCipher.Encrypt(handshake.RemoteNodeId, ackData.ReadAllBytesAsArray(), sizeBytes));
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network/Rlpx/Handshake/HandshakeService.cs:196]: In the end we have ReadAllBytesAsArray(), maybe we can do it with span?

File: src/Nethermind/Nethermind.Network.Discovery/NettyDiscoveryHandler.cs [hunk 7]
```diff
@@ -80,26 +80,24 @@ public override void ChannelReadComplete(IChannelHandlerContext context)
...
     public async void SendMsg(DiscoveryMsg discoveryMsg)
     {
-        byte[] msgBytes;
-
+        IByteBuffer msgBuffer;
         try
         {
             if (_logger.IsTrace) _logger.Trace($"Sending message: {discoveryMsg}");
-            msgBytes = Serialize(discoveryMsg);
+            msgBuffer = Serialize(discoveryMsg, ByteBufferAllocator.UnpooledByteBufferAllocator);
```
[the remark this answers is not in the document]
Comment (tanishqjasoria) [reply src/Nethermind/Nethermind.Network.Discovery/NettyDiscoveryHandler.cs:87]: yes, it is released in the WriteAndFlushAsync.
I used the same buffer that was used earlier.

File: src/Nethermind/Nethermind.Network.Discovery/Serializers/DiscoveryMsgSerializerBase.cs [hunk 8]
```diff
@@ -39,10 +41,11 @@ public abstract class DiscoveryMsgSerializerBase
...
     }
 
-    protected byte[] Serialize(byte type, Span<byte> data)
+    protected void Serialize(byte type, Span<byte> data, IByteBuffer byteBuffer)
     {
-        byte[] result = new byte[32 + 1 + data.Length + 64 + 1];
-        Span<byte> resultSpan = result.AsSpan();
+        int length = 32 + 1 + data.Length + 64 + 1;
+        byteBuffer.EnsureWritable(length);
+        Span<byte> resultSpan = stackalloc byte[length];
```
[the remark this answers is not in the document]
Comment (tanishqjasoria) [reply src/Nethermind/Nethermind.Network.Discovery/Serializers/DiscoveryMsgSerializerBase.cs:48]: yes, we can do that by moving WriterIndex. Will change this

File: src/Nethermind/Nethermind.Network/P2P/NettyRlpStream.cs (hunk 1 above)
Comment (tanishqjasoria) [reply src/Nethermind/Nethermind.Network/P2P/NettyRlpStream.cs:104]: I was under the impression Dispose is also called automatically?

**Comment by tanishqjasoria** [comment]
> Not all tests are passing. Flaky tests? How the PR was tested?

PR was tested syncing mainnet and analysing memory dump.
I thought the test was flaky but I think there is another issue, having a look

File: src/Nethermind/Nethermind.Network/P2P/NettyRlpStream.cs (hunk 1 above)
Comment (tanishqjasoria) [reply src/Nethermind/Nethermind.Network/P2P/NettyRlpStream.cs]: it turns out we dont need to call this after all

**Pull request closed without merging**